### PR TITLE
Fail build with HAL resource without serializer #11246

### PR DESCRIPTION
**Pull request by gytis**
https://github.com/quarkusio/quarkus/issues/11246

### CODE REVIEW

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/RestDataProcessor.java [hunk 1]
```diff
@@ -57,4 +63,36 @@ JsonbSerializerBuildItem registerJsonbSerializers() {
...
+        ResourcePropertiesAccessor propertiesAccessor = new ResourcePropertiesAccessor(index);
+        for (RestDataResourceBuildItem resource : resourceBuildItems) {
+            if (propertiesAccessor.isHal(resource.getResourceInfo().getClassInfo())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean hasJsonProcessor() {
```
Comment (gsmet) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/RestDataProcessor.java:85]: You should rely on the `Capabilities` here. The ones existing don't really cover your use case (I checked, even if their names sound appealing, they are not covering your needs) so you will need to add `RESTEASY_JACKSON` and `RESTEASY_JSONB` and produce them from the respective processors together with the others.

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/RestDataProcessor.java [hunk 2]
```diff
@@ -57,4 +63,36 @@ JsonbSerializerBuildItem registerJsonbSerializers() {
     RuntimeInitializedClassBuildItem el() {
         return new RuntimeInitializedClassBuildItem(EL.class.getCanonicalName());
     }
+
+    private void verifyHalResources(IndexView index, List<RestDataResourceBuildItem> resourceBuildItems) {
+        if (!hasJsonProcessor() && hasHalResource(index, resourceBuildItems)) {
+            logger.warn("Cannot generate HAL endpoints without a serializer. "
+                    + "Make sure you have either Jsonb or Jackson RESTEasy extension enabled.");
```
Comment (gsmet) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/RestDataProcessor.java:70]:
```suggestion
                    + "Make sure you have either RESTEasy JSON-B or Jackson extension enabled.");
```

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/RestDataProcessor.java [hunk 3]
```diff
@@ -57,4 +63,36 @@ JsonbSerializerBuildItem registerJsonbSerializers() {
...
+            throw new RestDataBuildException("Cannot generate HAL endpoints without a serializer");
```
Comment (gsmet) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/RestDataProcessor.java:71]: Either you log a warning or you throw an exception. In this case, I would just throw an exception with the proper message. 

Also I don't think you need a specific exception for this. An `IllegalStateException` would be perfectly adequate IMHO.

**Review by gsmet — DISMISSED** [review]
Added some comments inline. Feel free to ping me if something is unclear.

**Comment by gytis** [comment]
Thanks for the tip on capabilities, @gsmet. It makes a much more sense and keeps the code much cleaner. I changed the commit to use `RESTEASY_JSON` capability, I think that should be sufficient.

**Comment by gsmet** [comment]
Nope. That's why I was extremely clear in my comment. The `RESTEASY_JSON` capability is also produced for Vert.x JSON support and it won't work for you.

**Comment by gytis** [comment]
OK I'll add the new capabilities then.

**Comment by gytis** [comment]
@gsmet I've just noticed that there are two other capabilities: `REST_JACKSON` and `REST_JSONB`. They are produced by their respective server and client deployments both of which also bring in the required serialiser dependencies. I've just tried adding all of them one by one and serialisation worked correctly.
Of course, this could be a side effect which might go away in the future. I can still add new `RESTEASY_JACKSON` and `RESTEASY_JSONB` capabilities avoid that. But I just wanted to double check with you to avoid adding any unnecessary code.

**Comment by gsmet** [comment]
@gytis as I mentioned it twice now, I did the digging and you need new capabilities.

**Comment by gytis** [comment]
Sorry for asking multiple times. I just wanted to avoid adding unnecessary code.
I've added the new capabilities as you've said.

**Comment by gytis** [comment]
Is there something else missing in this PR?

**Review by geoand — APPROVED** [review]

**Pull request merged**